test/ReverseProxy.FunctionalTests/WebSocketTests.cs

```diff
@@ -138,10 +135,276 @@ public async Task FalseUpgradeTest()
+            client.Options.HttpVersion = HttpVersion.Version20;
+            client.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+            await SendWebSocketRequestAsync(client, uri, "HTTP/1.1", cts.Token);
+        }, cts.Token);
+    }
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public async Task WebSocket11_To_20(bool useHttps)
+    {
+        if (OperatingSystem.IsMacOS() && useHttps)
+        {
+            // Does not support ALPN until .NET 8
+            return;
+        }
+
+        using var cts = CreateTimer();
+
+        var test = CreateTestEnvironment();
+        test.ProxyProtocol = HttpProtocols.Http1;
+        test.DestinationProtocol = HttpProtocols.Http2;
+        test.DestionationHttpVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+        test.UseHttpsOnProxy = useHttps;
+        test.UseHttpsOnDestination = useHttps;
+
+        await test.Invoke(async uri =>
+        {
+            using var client = new ClientWebSocket();
+            client.Options.HttpVersion = HttpVersion.Version11;
+            client.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+            await SendWebSocketRequestAsync(client, uri, "HTTP/2", cts.Token);
+        }, cts.Token);
+    }
+
+    [Fact]
+    public async Task WebSocketFallbackFromH2()
+    {
+        using var cts = CreateTimer();
+
+        var test = CreateTestEnvironment();
+        test.ProxyProtocol = HttpProtocols.Http1;
+        // The destination doesn't support HTTP/2, as determined by ALPN
+        test.DestinationProtocol = HttpProtocols.Http1;
+        test.DestionationHttpVersion = HttpVersion.Version20;
+        test.UseHttpsOnDestination = true;
+
+        await test.Invoke(async uri =>
+        {
+            using var client = new ClientWebSocket();
+            await SendWebSocketRequestAsync(client, uri, "HTTP/1.1", cts.Token);
+        }, cts.Token);
+    }
+
+    // [Fact]
+    [Fact(Skip = "Manual test only, the CI doesn't always have the IIS Express test cert installed.")]
+    public async Task WebSocketFallbackFromH2WS()
+    {
+        if (!OperatingSystem.IsWindows())
+        {
+            // This test relies on Windows/HttpSys
+            return;
+        }
+
+        using var cts = CreateTimer();
+
+        var test = CreateTestEnvironment();
+        test.ProxyProtocol = HttpProtocols.Http1;
+        // The destination supports HTTP/2, but not H2WS
+        test.UseHttpSysOnDestination = true;
+        test.UseHttpsOnDestination = true;
+
+        await test.Invoke(async uri =>
+        {
+            using var client = new ClientWebSocket();
+            await SendWebSocketRequestAsync(client, uri, "HTTP/1.1", cts.Token);
+        }, cts.Token);
+    }
+
+    [Theory]
+    [InlineData(HttpVersionPolicy.RequestVersionExact, true)]
+    // [InlineData(HttpVersionPolicy.RequestVersionExact, false)] // Times out. https://github.com/dotnet/runtime/issues/80056
```

Comment:
```suggestion
#if NET8_0_OR_GREATER
    [InlineData(HttpVersionPolicy.RequestVersionExact, false)] // Times out. https://github.com/dotnet/runtime/issues/80056
#endif
```
I have a feeling we'll forget to re-enable it otherwise